compiler/annotation/type_seeder.rs

```diff
@@ -140,6 +140,28 @@ impl<'this, Snapshot: ReadableSnapshot> TypeGraphSeedingContext<'this, Snapshot>
         for nested_graph in nested_optionals {
             self.seed_types(nested_graph, context, vertices)?;
         }
+
+        // Prune abstract types from type annotations of thing variables
+        for annotated_vertex in vertices {
+            let (Vertex::Variable(id), annotations) = annotated_vertex else { continue; };
+            if self.variable_registry
+                .get_variable_category(*id)
+                .map_or(false, |cat| { cat.is_thing_category() }) {
+               let mut to_be_removed : Vec<Type> = Vec::new();
+               for annotated_type in annotations.iter() {
+                   let annotated_type_is_abstract = annotated_type
+                       .is_abstract(self.snapshot, self.type_manager)
+                       .map_err(|source| { TypeInferenceError::ConceptRead { source }})?;
+                   if annotated_type_is_abstract {
+                       to_be_removed.push(annotated_type.clone());
+                   }
+               }
+               for annotated_type in to_be_removed.iter() {
+                   annotations.remove(annotated_type);
+               }
```

Comment:
What we've done here is essentially a `try_retain` and I think all the 3 cases would be more readable if we defined that as a function. I think it's worth having atleast locally. It doesn't have to be generic:

```rust
fn try_retain<T: std::cmp::Ord + Clone, Err>(set: &mut BTreeSet<T>, fallible_check: impl Fn(&T) -> Result<bool, Err>) -> Result<(), Err>{
    let to_remove: Vec<T> = set.iter().filter_map(|item| {
        match fallible_check(item) {
            Err(err) => Some(Err(err)),
            Ok(false) => Some(Ok(item.clone())),
            Ok(true) => None,
        }
    }).collect::<Result<Vec<_>, _>>()?;
    to_remove.iter().for_each(|item| {set.remove(item);});
    Ok(())
}
```

Then, this would simply read:
```rust
 annotations.try_retain(|type_| {
  type_.is_abstract()).map_err(|source| { TypeInferenceError::ConceptRead { source }})?
});
```